cmd/argocd/commands/app.go

```diff
@@ -1799,13 +1821,51 @@ func NewApplicationWaitCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 }
 
 // printAppResources prints the resources of an application in a tabwriter table
-func printAppResources(w io.Writer, app *argoappv1.Application) {
+func printAppResources(w io.Writer, app *argoappv1.Application, discoveryClient discovery.DiscoveryInterface) {
 	_, _ = fmt.Fprintf(w, "GROUP\tKIND\tNAMESPACE\tNAME\tSTATUS\tHEALTH\tHOOK\tMESSAGE\n")
-	for _, res := range getResourceStates(app, nil) {
+	versionedResourceStates := getVersionedResourceStates(app, nil)
+	for _, res := range mergeDuplicateClusterLevelResources(versionedResourceStates, discoveryClient) {
 		_, _ = fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", res.Group, res.Kind, res.Namespace, res.Name, res.Status, res.Health, res.Hook, res.Message)
 	}
 }
 
+func mergeDuplicateClusterLevelResources(resourceStates []*versionedResourceState, discoveryClient discovery.DiscoveryInterface) []*versionedResourceState {
+	resourcesByKey := make(map[kube.ResourceKey]*versionedResourceState)
+	for _, resource := range resourceStates {
+		groupVersion := ""
+		if resource.Group != "" {
+			groupVersion += (resource.Group + "/")
+		}
+		groupVersion += resource.Version
+		apiResourceList, err := discoveryClient.ServerResourcesForGroupVersion(groupVersion)
+		if err != nil {
+			errors.CheckError(fmt.Errorf("unable to retrieve server resources for group and version combination %s: %w", groupVersion, err))
+		}
+
+		for _, apiResource := range apiResourceList.APIResources {
+			if resource.Group == apiResource.Group && resource.Version == apiResource.Version && resource.Kind == apiResource.Kind {
+				if !apiResource.Namespaced {
+					resource.Namespace = ""
```

Comment:
Can we update the backend to properly handle this case?

Reply:
I am unsure to what case are you refering. Could you please explain what you would like to have changed, and where?

Reply:
Essentially to not overrride the namespace in CLI, but make ArgoCD not set it in the first place for cluster resources, or unset it before returning data to CLI.

Reply:
That does make sense, and was part of my initial analysis. I will have to dig significantly deeper though, and it may take me up to next week ( since I am basically clueless about ArgoCD codebase ). My assumption currently is though that this happens in GitOps Engine library. I made a quite thorough analysis about the cause of the bug in the issue this PR belongs to.